lib/TPP/CombineXsmmPass.cpp

```diff
@@ -0,0 +1,179 @@
+//===CombineXsmmPass.cpp --------------------------------------*----C++-*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM
+// Exceptions. / See https://llvm.org/LICENSE.txt for license information. /
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+
+#include "TPP/BuilderUtils.h"
+#include "TPP/Dialect/Xsmm/XsmmOps.h"
+#include "TPP/Dialect/Xsmm/XsmmUtils.h"
+#include "TPP/Passes.h"
+#include "mlir/Dialect/Func/IR/FuncOps.h"
+#include "mlir/Dialect/SCF/IR/SCF.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+
+using namespace mlir;
+
+#include "TPP/Dialect/Xsmm/XsmmUtils.h"
+#include "TPP/VNNIUtils.h"
+#include <list>
+namespace mlir {
+namespace tpp {
+#define GEN_PASS_DEF_COMBINEXSMMOPPASS
+#include "TPP/Passes.h.inc"
+} // namespace tpp
+} // namespace mlir
+namespace {
+
+static FailureOr<DenseI64ArrayAttr>
+getSizesAndLeadingDimForBrgemmOp(RewriterBase &rewriter, xsmm::BrgemmOp opTy) {
+
+  auto memrefC = opTy.getOperand(3).getType();
+  auto memrefA = opTy.getOperand(1).getType();
+  auto memrefB = opTy.getOperand(2).getType();
+
+  int64_t m = memrefC.cast<ShapedType>().getShape()[0];
```

Comment:
This is unsafe. There's no guarantee that the type is indeed `ShapedType` and the cast will succeed. If it does not, a segfault will occur. Use the `isa` or `dyn_cast` patterns to make sure you're dealing with the right type.

Further unsafe for use of `MemRefType` casts below without checking it's a valid cast. Since XSMM only works on memrefs, just use `MemRefType` everywhere.